versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,180 @@
+import com.dremio.nessie.versioned.impl.condition.UpdateExpression;
+import com.dremio.nessie.versioned.store.Id;
+import com.dremio.nessie.versioned.store.LoadStep;
+import com.dremio.nessie.versioned.store.SaveOp;
+import com.dremio.nessie.versioned.store.Store;
+import com.dremio.nessie.versioned.store.ValueType;
+import com.dremio.nessie.versioned.store.mongodb.codecs.CodecProvider;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+
+/**
+ * This class implements the Store interface that is used by Nessie as a backing store for versioning of it's
+ * Git like behaviour.
+ * The MongoDbStore connects to an external MongoDB server.
+ */
+public class MongoDbStore implements Store {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbStore.class);
+
+  private final String databaseName;
+  private final MongoClientSettings mongoClientSettings;
+
+  // The client that connects to the MongoDB server.
+  protected MongoClient mongoClient;
+
+  // The database hosted by the MongoDB server.
+  private MongoDatabase mongoDatabase;
+  protected final Map<ValueType, MongoCollection<?>> collections;
+
+  /**
+   * Creates a store ready for connection to a MongoDB instance.
+   * @param connectionString the string to use to connect to MongoDB.
+   * @param databaseName the name of the database to retrieve
+   */
+  public MongoDbStore(ConnectionString connectionString, String databaseName) {
+    this.collections = new HashMap<>();
+    this.databaseName = databaseName;
+    final CodecRegistry codecRegistry = CodecRegistries.fromProviders(
+        new CodecProvider(),
+        PojoCodecProvider.builder().automatic(true).build(),
+        MongoClientSettings.getDefaultCodecRegistry());
+    this.mongoClientSettings = MongoClientSettings.builder()
+      .applyConnectionString(connectionString)
+      .codecRegistry(codecRegistry)
+      .build();
+  }
+
+  /**
+   * Gets a handle to an existing database or get a handle to a MongoDatabase instance if it does not exist. The new
+   * database will be lazily created.
+   * Since MongoDB creates databases and collections if they do not exist, there is no need to validate the presence of
+   * either before they are used. This creates or retrieves collections that map 1:1 to the enumerates in
+   * {@link com.dremio.nessie.versioned.store.ValueType}
+   */
+  @Override
+  public void start() {
+    mongoClient = MongoClients.create(mongoClientSettings);
+    mongoDatabase = mongoClient.getDatabase(databaseName);
+
+    // Initialise collections for each ValueType.
+    collections.put(ValueType.L1, mongoDatabase.getCollection(MongoDbConstants.L1_COLLECTION, L1.class));
```

Comment:
It seems like name and sharding should be configurable here.

Reply:
also optional (configurable) creation (as opposed to required). We already have a property for this, I believe.

Reply:
(these comments are about creation, realized that this is just capturing)

Reply:
What value would there be in the initialization flag, given that Mongo creates collections lazily if they aren't present already upon first use?

Reply:
Updating context - after discussion we'd want collection creation to be shard aware.

Reply:
Shard configuration should be via scripting outside of the application. Sharding is enabled via the admin database. It would make sense not to add admin commands to the application itself. All sharding configuration should be done in one place. The application does not need direct knowledge of sharding. Sharding will be deferred to after this PR.

Reply:
Added sharding to test environment using IndexManager as an example from Dremio.

Reply:
I'm pulling sharding of the test environment for now until we can fix it properly. Please review on the basis it is coming.